lib/filtergw/src/olsrd_filtergw.c

```diff
@@ -0,0 +1,307 @@
+#include "olsr_cfg.h"
+
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <sys/time.h>
+#include <net/route.h>
+#include <unistd.h>
+#include <errno.h>
+#include <time.h>
+
+static int is_allowlist = 0;
+
+static struct filter_list * add_to_filter_list(const char *, struct filter_list *);
+
+struct originator_list {
+  union olsr_ip_addr        originator;
+  struct originator_list *  next;
+};
+
+struct filter_group {
+  struct originator_list *  originator_list;
+  struct hna_group *        next;
+};
+
+static struct filter_group * filter_groups = NULL;
+
+/* -------------------------------------------------------------------------
+ * Function   : add_to_originator_list
+ * Description: Add a new ip to originator list
+ * Input      : originator_address - the address of originator
+ *              the_originator_list - the list of originators
+ * Output     : none
+ * Return     : a pointer to the newly added originator, i.e. start of the list
+ * Data Used  : none
+ * ------------------------------------------------------------------------- */
+/* add the valid IPs to the head of the list */
+static struct originator_list *
+add_to_originator_list(union olsr_ip_addr * originator_address, struct originator_list *the_originator_list)
+{
+  struct originator_list *new = calloc(1, sizeof(struct originator_list));
+  if (!new) {
+    olsr_exit("FILTERGW: Out of memory", EXIT_FAILURE);
+  }
+  memcpy(&new->originator, originator_address, olsr_cnf->ipsize);
+  new->next = the_originator_list;
+  return new;
+}
+
+/**
+ * read config file parameters
+ */
+static int
+set_plugin_filter(const char *value, void *data __attribute__ ((unused)), set_plugin_parameter_addon addon __attribute__ ((unused)))
+{
+  union olsr_ip_addr addr;
+
+  if (inet_pton(olsr_cnf->ip_version, value, &addr) <= 0) {
+    OLSR_PRINTF(0, "Illegal IP address \"%s\"", value);
+    return 1;
+  }
+
+  if (filter_groups == NULL) {
+      struct filter_group *new = calloc(1, sizeof(struct filter_group));
+      if (new == NULL) {
+        olsr_exit("FILTERGW: Out of memory", EXIT_FAILURE);
+      }
+      filter_groups = new;
+      new->next =  filter_groups;
+  }
+
+  filter_groups->originator_list = add_to_originator_list(&addr, filter_groups->originator_list);
+
+  return 0;
+}
+
+/* -------------------------------------------------------------------------
+ * Function   : olsr_drophna_parser
+ * Description: Function to be passed to the parser engine. This function
+ *              processes the incoming message and filters gw hna's.
+ * Input      : m      - message to parse
+ *              in_if  - interface to use (unused in this application)
+ *              ipaddr - IP-address to use (unused in this application)
+ * Output     : none
+ * Return     : false if message should be supressed, true otherwise
+ * Data Used  : none
+ * ------------------------------------------------------------------------- */
+bool
+olsrd_drophna_parser(
```

Comment:
you should rename this

Reply:
please make your comments here. https://github.com/OLSR/olsrd/pull/86

:) thanks I will change that.